test/cpp/tensorexpr/test_loopnest.cpp

```diff
@@ -123,6 +123,312 @@ void testExprSimple02() {
+  l.sliceTail(loops[0], 4, &head, &tail);
+
+  ASSERT_NE(head, nullptr);
+  ASSERT_NE(head, loops[0]);
+  ASSERT_NE(tail, nullptr);
+  ASSERT_NE(tail, loops[0]);
+
+  Block* body = getSimplifiedBody(l);
+  assertForRanges(body, {{0, 6}, {6, 10}});
+}
+
+void testExprSplitAndSlice() {
+  // 0: splitWithTail
+  // 1: sliceTail on inner loop
+  // 2: sliceHead on outer loop
+  KernelScope kernel_scope;
+  auto func = [](const ExprHandle& x) {
+    return ExprHandle(1.0f) + cast<float>(x);
+  };
+  Tensor* tensor = Compute("f", {{100, "x"}}, func);
+  LoopNest l({tensor});
+
+  For* outer;
+  For* inner;
+  For* tail;
+  std::vector<For*> loops = l.getLoopStmtsFor(tensor);
+  // outer: [0, 4)
+  // inner: [0, 21)
+  // tail:  [84, 100)
+  l.splitWithTail(loops[0], 21, &outer, &inner, &tail);
+
+  For* inner_head;
+  For* inner_tail;
+  l.sliceTail(inner, 2, &inner_head, &inner_tail);
+
+  For* outer_head;
+  For* outer_tail;
+  l.sliceHead(outer, 2, &outer_head, &outer_tail);
+
+  // for (int x_outer = 0; x_outer < 2; x_outer++) {
+  //   for (int x_inner = 0; x_inner < 19; x_inner++) {
+  //     f[21 * x_outer + x_inner] = 1.f + float(21 * x_outer + x_inner);
+  //   }
+  //   for (int x_inner = 19; x_inner < 21; x_inner++) {
+  //     f[21 * x_outer + x_inner] = 1.f + float(21 * x_outer + x_inner);
+  //   }
+  // }
+  // for (int x_outer = 2; x_outer < 4; x_outer++) {
+  //   for (int x_inner = 0; x_inner < 19; x_inner++) {
+  //     f[21 * x_outer + x_inner] = 1.f + float(21 * x_outer + x_inner);
+  //   }
+  //   for (int x_inner = 19; x_inner < 21; x_inner++) {
+  //     f[21 * x_outer + x_inner] = 1.f + float(21 * x_outer + x_inner);
+  //   }
+  // }
+  // for (int x_tail = 0; x_tail < 16; x_tail++) {
+  //   f[x_tail + 84] = 1.f + float(x_tail + 84);
+  // }
+  Block* body = getSimplifiedBody(l);
+  assertForRanges(body, {{0, 2}, {2, 4}, {0, 16}});
+
+  auto biter = body->begin();
+
+  For* loop = dynamic_cast<For*>(*biter++);
+  assertForRanges(loop->body(), {{0, 19}, {19, 21}});
+
+  loop = dynamic_cast<For*>(*biter);
+  assertForRanges(loop->body(), {{0, 19}, {19, 21}});
+}
+
+void testExprSliceAndNormalize() {
+  // 0: sliceHead
+  // 1: normalize tail
+  KernelScope kernel_scope;
+  auto func = [](const ExprHandle& x) {
+    return ExprHandle(1.0f) + cast<float>(x);
+  };
+  Tensor* tensor = Compute("f", {{10, "x"}}, func);
+  LoopNest l({tensor});
+  std::vector<For*> loops = l.getLoopStmtsFor(tensor);
+
+  For* head;
+  For* tail;
+  l.sliceHead(loops[0], 2, &head, &tail);
+  // head: [0, 2)
+  // tail: [2, 10)
+
+  For* normalized_tail;
+  LoopNest::normalize(tail, &normalized_tail, false);
+  // normalized_tail: [0, 8)
+
+  Block* body = getSimplifiedBody(l);
+  assertForRanges(body, {{0, 2}, {0, 8}});
+}
+
+template <typename T>
+T evalExpr(const ExprHandle& expr, const VarHandle& var, T value) {
+  ExprEval<SimpleIREvaluator> eval(expr, {var});
+  return eval.value<T>(value);
+}
+
+void testExprSliceWithVariableDimension() {
```

Comment:
This is a good test, but it probably would have been simplier via FileCheck rather than binding the variable,

Reply:
The simplified IR would be like 
```
{
  for (int x = 0; x < Min(2, dim, 1); x++) {
    f[x] = x;
  }
  for (int x = Min(2, dim, 1); x < Min((Min(2, dim, 1)) + (Max(0, (dim - (Min(2, dim, 1))) - 2, 1)), dim, 1); x++) {
    f[x] = x;
  }
  for (int x = Min((Min(2, dim, 1)) + (Max(0, (dim - (Min(2, dim, 1))) - 2, 1)), dim, 1); x < dim; x++) {
    f[x] = x;
  }
}
```

Which is hard to read, binding variable makes the check easier to understand. WDYT?

Reply:
@navahgar is working on simplifying min and max patterns but I take you point. I'm fine with how it is.

Reply:
Yeah. Here is the PR that simplifies some of these Min-Max patterns: https://github.com/pytorch/pytorch/pull/44142